### DISCUSSION TITLE
How to save images?

### DISCUSSION THREAD

**Discussion by aurelienbl** [colour-science/colour#736] [Q&A] [answered]
Hello,

Thank you very much for this great package. 
I have a very simple question. Probably to simple that I could not find the solution in the documentation.

How can I save image on disk in a proper format (png, jpg...). 
More precisely, I have followed the colour-correction tutorial, and it is not clear to me how you save the corrected image in the end.

A bit of help would be appreciated!

Best,
Aurélien

**Comment by thomasmansencal** [comment]
Hi @aurelienbl,

Simply passing the `filename=my_image_name.png` kwarg to most of the plotting definitions should be enough!

Cheers,

Thomas

**Comment by aurelienbl** [comment]
Hello Thomas,

Thank you for your replied, but I still have inconsistent result. 

After the correction is done, I do:
`colour.io.write_image(corrected_image, 'corrected_image.png')`

Here is the result I have:
![failed_saved](https://user-images.githubusercontent.com/5089617/74274844-15949d80-4d13-11ea-84d3-e3d47e00425d.jpeg)
On the left, the corrected as displayed by the tutorial within the python interpreter. On the right the saved image. 
I don't doubt that I have something wrong, but I don't see what!

Cheers,
Aurélien

**Comment by thomasmansencal** [comment]
Oh sorry, I did not read carefully and you are not using the plotting code.

In order to write a PNG or JPG, i.e. 8-bit data, image to disk, you would need to non-linearly encode the image first, e.g. `my_image = colour.cctf_encoding(my_image)` then you can write it.

Is it possible to let me know which image IO backend you are using by giving me the output of the following definition: `colour.utilities.describe_environment()`

Cheers,

Thomas

**Comment by aurelienbl** [comment]
Hum I think I tried that, Here is corrected png.
![corrected](https://user-images.githubusercontent.com/5089617/74277329-a0779700-4d17-11ea-9944-543b29ac81ef.png)
(for reference I did 
`colour.io.write_image(colour.cctf_encoding(corrected) ,'corrected.png')`) 

Here is the output of  `colour.utilities.describe_environment()`:
`defaultdict(collections.OrderedDict,
            {'Interpreter': OrderedDict([('python',
                           '3.6.5 (default, Apr 25 2018, 14:23:58) \n[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.1)]')]),
             'colour-science.org': OrderedDict([('colour', '0.3.15'),
                          ('colour-checker-detection', 'v0.1.0-56-g246e441')]),
             'Runtime': OrderedDict([('imageio', '2.5.0'),
                          ('matplotlib', '3.1.1'),
                          ('networkx', '2.2'),
                          ('numpy', '1.16.4'),
                          ('pandas', '0.25.3'),
                          ('scipy', '1.1.0'),
                          ('six', '1.11.0'),
                          ('opencv', '4.2.0')])})`


Thanks for the help!

**Comment by thomasmansencal** [comment]
Looks like the image is compressed somehow! Let me get back to you tonight.

Bests,

Thomas

**Comment by thomasmansencal** [comment, marked as answer]
Allright, took 5 minutes to look at that, it seems like `imageio` (PIL?) is trying to be smart and whenever it has floating point data casted to 8-bit, it remaps it as follows: [min, max] --> [0, 255] thus you should do something like that:

`my_image = (colour.cctf_encoding(np.clip(my_image, 0, 1)) * 255).astype(np.uint8)`

**Comment by aurelienbl** [comment]
Great ! It works fine like this. 
Thank tou so much for the help.

Aurélien

**Comment by thomasmansencal** [comment]
You are welcome! :)
